File: 36. Valid Sudoku/isValidSudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
const isValidSudoku = (board) => {
  const rows = new Map();
  const cols = new Map();
  const squars = new Map();

  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      let num = board[row][col];
      if (num === '.') {
        continue;
      }

      if (!rows.has(row)) {
        rows.set(row, new Set());
      }

      if (!cols.has(col)) {
        cols.set(col, new Set());
      }

      let squareKey = `${Math.floor(row / 3)}-${Math.floor(col / 3)}`;

      if (!squars.has(squareKey)) {
        squars.set(squareKey, new Set());
      }

      let isInRows = rows.get(row).has(num);
      let isInCols = cols.get(col).has(num);
      let isInSquare = squars.get(squareKey).has(num);

      if (isInRows || isInCols || isInSquare) {
        return false;
      }

      rows.get(row).add(num)
      cols.get(col).add(num);
      squars.get(squareKey).add(num);
    }
  }
  
  return true;
};

const board = [
   ["5","3",".",".","7",".",".",".","."]
  ,["6",".",".","1","9","5",".",".","."]
  ,[".","9","8",".",".",".",".","6","."]
  ,["8",".",".",".","6",".",".",".","3"]
  ,["4",".",".","8",".","3",".",".","1"]
  ,["7",".",".",".","2",".",".",".","6"]
  ,[".","6",".",".",".",".","2","8","."]
  ,[".",".",".","4","1","9",".",".","5"]
  ,[".",".",".",".","8",".",".","7","9"]
];

console.log('isValidSudoku: ', isValidSudoku(board));

File: 36. Valid Sudoku/isValidSudoku.py
from typing import List
from collections import defaultdict

class Solution:
  def isValidSudoku(self, board: List[List[str]]) -> bool:
    cols = defaultdict(set)
    rows = defaultdict(set)
    squares = defaultdict(set)

    for row in range(9):
      for col in range(9):
        num = board[row][col]
        if num == '.':
          continue
        
        if (num in rows[row] or
          num in cols[col] or
          num in squares[(row // 3, col // 3)]):
          return False

        rows[row].add(num)
        cols[col].add(num)
        squares[(row // 3, col // 3)].add(num)

    return True
  
solution = Solution()
board = [
  ["5","3",".",".","7",".",".",".","."]
 ,["6",".",".","1","9","5",".",".","."]
 ,[".","9","8",".",".",".",".","6","."]
 ,["8",".",".",".","6",".",".",".","3"]
 ,["4",".",".","8",".","3",".",".","1"]
 ,["7",".",".",".","2",".",".",".","6"]
 ,[".","6",".",".",".",".","2","8","."]
 ,[".",".",".","4","1","9",".",".","5"]
 ,[".",".",".",".","8",".",".","7","9"]
]
print('isValidSudoku: ', solution.isValidSudoku(board))
